### PR TITLE
Reservation

### PR DESCRIPTION
**Pull request by bernard2b**

### CODE REVIEW

File: src/main/java/com/example/dr_app/controller/ReservationController.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ReservationController {
+
+    @Autowired
+    ReservationService reservationService;
+
+    @PostMapping("/makeReservation")
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/controller/ReservationController.java]: Can you change the name to create-reservation ? Also please change the api endpoint names to kebab case?

File: src/main/java/com/example/dr_app/controller/ReservationController.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    public ResponseEntity<Reservation> makeReservation(@RequestParam Long doctorId, @RequestParam Long patientId) {
+        try {
+            Reservation savedReservation = reservationService.saveReservationData(doctorId, patientId);
+            return ResponseEntity.ok(savedReservation);
+        } catch (IllegalStateException e) {
+            return ResponseEntity.badRequest().body(null);
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/controller/ReservationController.java]: Maybe you can send back the failed entity with the body. In general it's not a good idea to send back a null body.

File: src/main/java/com/example/dr_app/model/Patient.java [hunk 3]
```diff
@@ -58,5 +58,10 @@ public class Patient {
             unique = true
     )
     private String email;
-    private Long doctor_id;
+
+    private Long dr_patient_fk;
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/model/Patient.java]: why do you use a snake case here? I would just go with the simple camel case.

File: src/main/java/com/example/dr_app/model/Reservation.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(
+            name = "id",
+            updatable = false
+    )
+    private Long id;
+    private Long doctorId;
+    private Long patientId;
+    @Temporal(TemporalType.TIMESTAMP)
+    private LocalDateTime localDateTime;
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/model/Reservation.java]: Woud be better to use a date for this variable name. Also, you need to convert any local date to UTC before saving it to the db.

File: src/main/java/com/example/dr_app/service/ReservationService.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    ReservationRepository reservationRepository;
+    @Autowired
+    DoctorRepository doctorRepository;
+    @Autowired
+    PatientRepository patientRepository;
+
+    public Reservation saveReservationData(Long doctorId, Long patientId) {
+        
+        Doctor doctor = doctorRepository.findById(doctorId).orElseThrow(() ->
+                new ResponseStatusException(HttpStatus.BAD_REQUEST, "No doctor found"));
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/service/ReservationService.java]: Please make this error response more descriptive, with info about the doctor/patient.

File: src/main/java/com/example/dr_app/service/ReservationService.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        Patient patient = patientRepository.findById(patientId).orElseThrow(() ->
+                new ResponseStatusException(HttpStatus.BAD_REQUEST, "No patient found"));
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/service/ReservationService.java]: Same as above

File: src/main/java/com/example/dr_app/service/ReservationService.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+        Reservation reservation = new Reservation();
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/service/ReservationService.java]: can you use a builder instead of the setters?

**Review by Alex880627 — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/dr_app/controller/ReservationController.java (hunk 1 above)
Comment (bernard2b) [reply src/main/java/com/example/dr_app/controller/ReservationController.java]: All the endpoints should be in kebab case?

Comment (Alex880627) [reply src/main/java/com/example/dr_app/controller/ReservationController.java]: yes, sorry for late reply

File: src/main/java/com/example/dr_app/model/Reservation.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    private Date reservationDate;
+    private Time reservationTime;
+
+    public Reservation() {
+        // Default constructor
+    }
+
+    public void setDate(LocalDateTime date) {
+        // Convert LocalDateTime to UTC
+        ZonedDateTime utcDateTime = date.atZone(ZoneOffset.UTC);
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/model/Reservation.java:50]: this makes no sense, you convert it to UTC and after back to the local date. Should be saved in UTC

File: src/main/java/com/example/dr_app/controller/ReservationController.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    @Autowired
+    ReservationService reservationService;
+
+    @PostMapping("/create-reservation")
+    public ResponseEntity<Reservation> makeReservation(@RequestParam Long doctorId, @RequestParam Long patientId) {
+        try {
+            Reservation savedReservation = reservationService.saveReservationData(doctorId, patientId);
+            return ResponseEntity.ok(savedReservation);
+        } catch (IllegalStateException e) {
+            Reservation failedreservation = new Reservation();
```
Comment (Alex880627) [inline src/main/java/com/example/dr_app/controller/ReservationController.java:23]: this shouldn't be a new Reservation, just an error with the info's about the failed reservation. Maybe I wasn't clear enough. Sorry.

**Pull request merged**